subtest/contributivity.py

```diff
@@ -938,6 +938,103 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
+            .dot(np.nan_to_num(relative_perf_matrix))
+
+        # Return contributivity scores
+        self.name = "Federated step by step quadratic scores"
+        self.contributivity_scores = contributivity_scores
+        self.normalized_scores = self.contributivity_scores / np.sum(
+            self.contributivity_scores
+        )
+        end = timer()
+        self.computation_time_sec = end - start
+
+    def federated_SBS_constant(self, the_scenario):
+        start = timer()
+        logger.info(
+            "# Launching computation of perf. scores of constant"
+            "performance increase compared to previous collective model")
+
+        relative_perf_matrix = self.compute_relative_perf_matrix(the_scenario)
+
+        # Calculate contributivity score as average of the relative performance for each round for each partner
+        contributivity_scores = np.nanmean(relative_perf_matrix, axis=0)
+
+        # Return contributivity scores
+        self.name = "Federated step by step constant scores"
+        self.contributivity_scores = contributivity_scores
+        self.normalized_scores = self.contributivity_scores / np.sum(
+            self.contributivity_scores
+        )
+        end = timer()
+        self.computation_time_sec = end - start
+
+    def compute_relative_perf_matrix(self, the_scenario):
+
+        # Define proportion of initial and final computation rounds to skip (default 10% each)
+        init_comp_rounds_skipped = 0.1
+        final_comp_rounds_skipped = 0.1
+
+        # Fetch score matrices from computation and scenario characteristics
+        multi_partner_learning = the_scenario.mpl
+        score_matrix_collective_models = multi_partner_learning.score_matrix_collective_models[:, 1:]
+        score_matrix_per_partner = multi_partner_learning.score_matrix_per_partner
+        partners_count = multi_partner_learning.partners_count
+        epoch_count = multi_partner_learning.epoch_count
+        minibatch_count = multi_partner_learning.minibatch_count
+
+        # Calculate first and last computation round kept for contributivity measure
+        first_comp_round_kept = int(np.round(epoch_count * minibatch_count * init_comp_rounds_skipped))
+        last_comp_round_kept = int(np.round(epoch_count * minibatch_count * (1 - final_comp_rounds_skipped)))
+
+        # Reshape scores matrices
+        scores_matrix_collective_reshape = np.reshape(score_matrix_collective_models,
+                                                      (epoch_count * minibatch_count))
+        score_matrix_per_partner_reshape = np.reshape(score_matrix_per_partner,
+                                                      (epoch_count * minibatch_count, partners_count))
+
+        # Calculate relative performance matrix
+        score_matrix_performance_rel = np.divide(score_matrix_per_partner_reshape,
+                                                 scores_matrix_collective_reshape[:, None])
```

Comment:
I might have misunderstood the method we wanted to implement, but I expected that here we would compute _the increment_ of the score for a partner on one minibatch step divided by _the increment_ of the collective score on one minibatch step. Here it seems that you manipulate the score at the end of each step and not its increase on each step. Was it your intention or is it a mistake?

Reply:
@Thomas-Galtier see https://github.com/SubstraFoundation/distributed-learning-contributivity/issues/105#issuecomment-648008117
After some tests it appeared that absolute increments were not optimal because the model could "unlearn" for each of the partners (meaning individual scores of each partner would be lower than the score of the aggregated model at previous step)

To solve the issue of regularization coming from this situation with @RomainGoussault , we had proposed to compute relative score increases (or decreases) rather than absolute, which is what is implemented here -- maybe this needs to be clarified in the issue as the method has slightly changed?